src/namedtuples.jl

```diff
@@ -1,54 +1,112 @@
 Base.size(nt::NamedTupleIterator) = (length(nt.x),)
 
-function Base.iterate(rows::NamedTupleIterator{Schema{names, T}}, st=()) where {names, T}
+@inline function Base.iterate(rows::NamedTupleIterator{Schema{names, T}, T1, T2}) where {names, T, T1, T2}
+    if @generated
+        vals = Tuple(:(getcolumn(row, $(fieldtype(T, i)), $i, $(quot(names[i])))) for i = 1:fieldcount(T))
+        return quote
+            rows.st === nothing && return nothing
+            row, st = rows.st
+            return $(NamedTuple{Base.map(Symbol, names), T})(($(vals...),)), st
+        end
+    else
+        rows.st === nothing && return nothing
+        row, st = rows.st
+        return NamedTuple{Base.map(Symbol, names), T}(Tuple(getcolumn(row, fieldtype(T, i), i, names[i]) for i = 1:fieldcount(T))), st
+    end
+end
+
+@inline function Base.iterate(rows::NamedTupleIterator{Schema{names, T}}, st) where {names, T}
     if @generated
-        vals = Tuple(:(getproperty(row, $(fieldtype(T, i)), $i, $(quot(names[i])))) for i = 1:fieldcount(T))
+        vals = Tuple(:(getcolumn(row, $(fieldtype(T, i)), $i, $(quot(names[i])))) for i = 1:fieldcount(T))
         return quote
-            x = iterate(rows.x, st...)
+            x = iterate(rows.x, st)
             x === nothing && return nothing
             row, st = x
-            return $(NamedTuple{Base.map(Symbol, names), T})(($(vals...),)), (st,)
+            return $(NamedTuple{Base.map(Symbol, names), T})(($(vals...),)), st
         end
     else
-        x = iterate(rows.x, st...)
+        x = iterate(rows.x, st)
         x === nothing && return nothing
         row, st = x
-        return NamedTuple{Base.map(Symbol, names), T}(Tuple(getproperty(row, fieldtype(T, i), i, names[i]) for i = 1:fieldcount(T))), (st,)
+        return NamedTuple{Base.map(Symbol, names), T}(Tuple(getcolumn(row, fieldtype(T, i), i, names[i]) for i = 1:fieldcount(T))), st
     end
 end
 
-# unknown schema case
-function Base.iterate(rows::NamedTupleIterator{Nothing, T}, st=()) where {T}
-    x = iterate(rows.x, st...)
+@inline function Base.iterate(rows::NamedTupleIterator{Schema{names, nothing}}) where {names}
+    rows.st === nothing && return nothing
+    row, st = rows.st
+    return NamedTuple{names}(Tuple(getcolumn(row, nm) for nm in names)), st
+end
+
+@inline function Base.iterate(rows::NamedTupleIterator{Schema{names, nothing}}, st) where {names}
+    x = iterate(rows.x, st)
     x === nothing && return nothing
     row, st = x
-    names = Tuple(propertynames(row))
-    return NamedTuple{Base.map(Symbol, names)}(Tuple(getproperty(row, nm) for nm in names)), (st,)
+    return NamedTuple{names}(Tuple(getcolumn(row, nm) for nm in names)), st
 end
 
-namedtupleiterator(::Type{T}, rows::S) where {T <: NamedTuple, S} = rows
-namedtupleiterator(::Type{T}, rows::S) where {T, S} = NamedTupleIterator{typeof(schema(rows)), S}(rows)
-
 # sink function
+"""
+    Tables.rowtable(x) => Vector{NamedTuple}
+    Tables.rowtable(rt, x) => rt
```

Comment:
I don't know enough context to understand why `rowtable(rt, x)` and `columntable(ct, x)` exist. But wouldn't it be better to introduce `tablecat(table, tables...)` or maybe even `Tables.cat(table, tables...)`?

Reply:
Yeah, we should probably remove the 2 appending methods; I agree they don't really make sense.